clvm-traits/src/curried_args.rs

```diff
@@ -0,0 +1,86 @@
+use clvmr::{
+    allocator::{NodePtr, SExp},
+    Allocator,
+};
+
+use crate::{
+    clvm_list, clvm_quote, destructure_list, destructure_quote, match_list, match_quote, Error,
+    FromClvm, MatchByte, Result, ToClvm,
+};
+
+#[derive(Debug, Clone)]
+pub struct CurriedArgs(pub Vec<NodePtr>);
+
+impl FromClvm for CurriedArgs {
+    fn from_clvm(a: &Allocator, mut ptr: NodePtr) -> Result<Self> {
+        let mut items = Vec::new();
+        loop {
+            match a.sexp(ptr) {
+                SExp::Atom => {
+                    if ptr == a.one() {
+                        return Ok(Self(items));
+                    } else {
+                        return Err(Error::ExpectedOne(ptr));
+                    }
+                }
+                SExp::Pair(..) => {
+                    let destructure_list!(_, destructure_quote!(first), rest) =
+                        <match_list!(MatchByte<4>, match_quote!(NodePtr), NodePtr)>::from_clvm(
+                            a, ptr,
+                        )?;
+
+                    items.push(first);
+                    ptr = rest;
+                }
+            }
+        }
+    }
+}
+
+impl ToClvm for CurriedArgs {
+    fn to_clvm(&self, a: &mut Allocator) -> Result<NodePtr> {
+        let mut result = a.one();
+        for item in self.0.iter().rev() {
+            result = clvm_list!(4, clvm_quote!(item), result).to_clvm(a)?;
+        }
+        Ok(result)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use clvmr::serde::node_to_bytes;
+
+    use super::*;
+
+    fn check(a: &mut Allocator, args: CurriedArgs, expected: &str) {
+        let ptr = args.to_clvm(a).unwrap();
+        let actual = node_to_bytes(a, ptr).unwrap();
+        assert_eq!(hex::encode(actual), expected);
+
+        let roundtrip = CurriedArgs::from_clvm(a, ptr).unwrap();
+
+        for (i, arg) in roundtrip.0.iter().enumerate() {
+            let actual = node_to_bytes(a, *arg).unwrap();
+            let expected = node_to_bytes(a, args.0[i]).unwrap();
+            assert_eq!(hex::encode(actual), hex::encode(expected));
+        }
+    }
+
+    #[test]
+    fn curried_args() {
+        let a = &mut Allocator::new();
+
+        let args = CurriedArgs(vec![
+            "a".to_clvm(a).unwrap(),
+            "b".to_clvm(a).unwrap(),
+            "c".to_clvm(a).unwrap(),
+        ]);
+
+        check(
+            a,
+            args,
+            "ff04ffff0161ffff04ffff0162ffff04ffff0163ff01808080",
+        );
+    }
```

Comment:
I would like to see some edge cases as well. like, 0 curried args and failure to match

Reply:
I think it would be good to have a test covering this failure case as well:

https://coveralls.io/builds/62027638/source?filename=clvm-traits%2Fsrc%2Fcurried_args.rs#L23

Reply:
since this function will take untrusted input, I think it's especially important to be very confident in it not breaking. We add tests to allow future changes, refactoring and optimizations, while being confident that we're not changing behavior